File: app/src/main/java/org/ole/planet/myplanet/base/BaseContainerFragment.java
package org.ole.planet.myplanet.base;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.AppCompatRatingBar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.google.gson.JsonObject;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.callback.OnHomeItemClickListener;
import org.ole.planet.myplanet.model.RealmMyLibrary;
import org.ole.planet.myplanet.service.UserProfileDbHandler;
import org.ole.planet.myplanet.ui.course.AdapterCourses;
import org.ole.planet.myplanet.ui.viewer.AudioPlayerActivity;
import org.ole.planet.myplanet.ui.viewer.CSVViewerActivity;
import org.ole.planet.myplanet.ui.viewer.ImageViewerActivity;
import org.ole.planet.myplanet.ui.viewer.MarkdownViewerActivity;
import org.ole.planet.myplanet.ui.viewer.PDFReaderActivity;
import org.ole.planet.myplanet.ui.viewer.TextFileViewerActivity;
import org.ole.planet.myplanet.ui.viewer.VideoPlayerActivity;
import org.ole.planet.myplanet.utilities.FileUtils;
import org.ole.planet.myplanet.utilities.Utilities;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import io.realm.RealmResults;

public abstract class BaseContainerFragment extends BaseResourceFragment {
    public TextView timesRated, rating;
    public AppCompatRatingBar ratingBar;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        profileDbHandler = new UserProfileDbHandler(getActivity());
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    public void setRatings(JsonObject object) {
        if (object != null) {
            AdapterCourses.showRating(object, rating, timesRated, ratingBar);
        }
    }

    public void initRatingView(View v) {
        timesRated = v.findViewById(R.id.times_rated);
        rating = v.findViewById(R.id.tv_rating);
        ratingBar = v.findViewById(R.id.rating_bar);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnHomeItemClickListener) {
            homeItemClickListener = (OnHomeItemClickListener) context;
        }
    }

    public void openIntent(RealmMyLibrary items, Class typeClass) {
        Intent fileOpenIntent = new Intent(getActivity(), typeClass);
        fileOpenIntent.putExtra("TOUCHED_FILE", items.getId() + "/" + items.getResourceLocalAddress());
        startActivity(fileOpenIntent);
    }


    public void openResource(RealmMyLibrary items) {
        if (items.getResourceOffline() != null && items.isResourceOffline()) {
            openFileType(items, "offline");
        } else if (FileUtils.getFileExtension(items.getResourceLocalAddress()).equals("mp4")) {
            openFileType(items, "online");
        } else {
            ArrayList<String> arrayList = new ArrayList<>();
            arrayList.add(Utilities.getUrl(items, settings));
            startDownload(arrayList);
        }
    }


    public void checkFileExtension(RealmMyLibrary items) {
        String filenameArray[] = items.getResourceLocalAddress().split("\\.");
        String extension = filenameArray[filenameArray.length - 1];
        switch (extension) {
            case "pdf":
                openIntent(items, PDFReaderActivity.class);
                break;
            case "bmp":
            case "gif":
            case "jpg":
            case "png":
            case "webp":
                openIntent(items, ImageViewerActivity.class);
                break;
            case "mp4":
                playVideo("offline", items);
                break;
            default:
                checkMoreFileExtensions(extension, items);
                break;
        }
    }

    public void checkMoreFileExtensions(String extension, RealmMyLibrary items) {
        switch (extension) {
            case "txt":
                openIntent(items, TextFileViewerActivity.class);
                break;
            case "md":
                openIntent(items, MarkdownViewerActivity.class);
                break;
            case "csv":
                openIntent(items, CSVViewerActivity.class);
                break;
            case "mp3":
                openIntent(items, AudioPlayerActivity.class);
                break;
            default:
                Toast.makeText(getActivity(), "This file type is currently unsupported", Toast.LENGTH_LONG).show();
                break;
        }

    }

    public void openFileType(final RealmMyLibrary items, String videotype) {
        if (FileUtils.getFileExtension(items.getResourceLocalAddress()).equals("mp4")) {
            playVideo(videotype, items);
        } else {
            checkFileExtension(items);
        }
    }


    public void playVideo(String videoType, final RealmMyLibrary items) {
        Intent intent = new Intent(getActivity(), VideoPlayerActivity.class);
        Bundle bundle = new Bundle();
        bundle.putString("videoType", videoType);
        if (videoType.equals("online")) {
            bundle.putString("videoURL", "" + Utilities.getUrl(items, settings));
            Log.e("AUTH", "" + auth);
            bundle.putString("Auth", "" + auth);
        } else if (videoType.equals("offline")) {
            bundle.putString("videoURL", "" + Uri.fromFile(new File("" + FileUtils.getSDPathFromUrl(items.getResourceRemoteAddress()))));
            bundle.putString("Auth", "");
        }
        intent.putExtras(bundle);
        startActivity(intent);
    }

    public void showResourceList(List<RealmMyLibrary> downloadedResources) {

        AlertDialog.Builder builderSingle = new AlertDialog.Builder(getActivity());
        builderSingle.setTitle("Select resource to open : ");

        final ArrayAdapter<RealmMyLibrary> arrayAdapter = new ArrayAdapter<RealmMyLibrary>(getActivity(), android.R.layout.select_dialog_item, downloadedResources) {
            @NonNull
            @Override
            public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                if (convertView == null)
                    convertView = LayoutInflater.from(getActivity()).inflate(android.R.layout.select_dialog_item, parent, false);
                TextView tv = (TextView) convertView;
                RealmMyLibrary library = getItem(position);
                tv.setCompoundDrawablesWithIntrinsicBounds(0,0,(library.isResourceOffline()? R.drawable.ic_eye : R.drawable.ic_download),  0);
                tv.setText(library.getTitle());
                return tv;
            }
        };
        builderSingle.setAdapter(arrayAdapter, (dialogInterface, i) -> {
            RealmMyLibrary library = arrayAdapter.getItem(i);
            openResource(library);
        });
        builderSingle.setNegativeButton("Dismiss", null).show();

    }

    public void setOpenResourceButton(List<RealmMyLibrary> downloadedResources, Button btnOpen) {
        if (downloadedResources == null || downloadedResources.size() == 0) {
            btnOpen.setVisibility(View.GONE);
        } else {
            btnOpen.setVisibility(View.VISIBLE);
            btnOpen.setOnClickListener(view -> {
                if (downloadedResources.size() == 1) {
                    openResource(downloadedResources.get(0));
                } else {
                    showResourceList(downloadedResources);
                }
            });
        }
    }

    public void setResourceButton(RealmResults resources, Button btnResources) {
        if (resources == null || resources.size() == 0) {
            btnResources.setVisibility(View.GONE);
        } else {
            btnResources.setVisibility(View.VISIBLE);
            btnResources.setText("Resources [" + resources.size() + "]");
            btnResources.setOnClickListener(view -> {
                if (resources.size() > 0)
                    showDownloadDialog(resources);
            });
        }

    }

}

File: app/src/main/java/org/ole/planet/myplanet/base/BaseContainerFragment.kt
package org.ole.planet.myplanet.base

import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.TextView
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.view.ContextThemeWrapper
import androidx.appcompat.widget.AppCompatRatingBar
import androidx.core.content.FileProvider
import androidx.core.net.toUri
import androidx.lifecycle.lifecycleScope
import com.google.gson.JsonObject
import dagger.hilt.android.AndroidEntryPoint
import java.io.File
import kotlinx.coroutines.launch
import org.ole.planet.myplanet.BuildConfig
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.base.BasePermissionActivity.Companion.hasInstallPermission
import org.ole.planet.myplanet.callback.OnHomeItemClickListener
import org.ole.planet.myplanet.callback.OnRatingChangeListener
import org.ole.planet.myplanet.model.RealmMyLibrary
import org.ole.planet.myplanet.service.UserProfileDbHandler.Companion.KEY_RESOURCE_DOWNLOAD
import org.ole.planet.myplanet.utilities.NavigationHelper
import org.ole.planet.myplanet.ui.viewer.WebViewActivity
import org.ole.planet.myplanet.utilities.CourseRatingUtils
import org.ole.planet.myplanet.utilities.FileUtils
import org.ole.planet.myplanet.utilities.ResourceOpener
import org.ole.planet.myplanet.utilities.SharedPrefManager
import org.ole.planet.myplanet.utilities.UrlUtils
import org.ole.planet.myplanet.utilities.Utilities

@AndroidEntryPoint
abstract class BaseContainerFragment : BaseResourceFragment() {
    private var timesRated: TextView? = null
    var rating: TextView? = null
    private var ratingBar: AppCompatRatingBar? = null
    private val installUnknownSourcesRequestCode = 112
    private var hasInstallPermissionValue = false
    private var currentLibrary: RealmMyLibrary? = null
    private var installApkLauncher: ActivityResultLauncher<Intent>? = null
    lateinit var prefData: SharedPrefManager
    private var pendingAutoOpenLibrary: RealmMyLibrary? = null
    private var shouldAutoOpenAfterDownload = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        hasInstallPermissionValue = hasInstallPermission(requireContext())
        if (!BuildConfig.LITE) {
            installApkLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
                if (result.resultCode == Activity.RESULT_OK) {
                    currentLibrary?.let {
                        installApk(it)
                        currentLibrary = null
                    }
                }
            }
        }
        prefData = SharedPrefManager(requireContext())
    }

    fun setRatings(`object`: JsonObject?) {
        if (`object` != null) {
            CourseRatingUtils.showRating(requireContext(), `object`, rating, timesRated, ratingBar)
        }
    }
    fun getUrlsAndStartDownload(lib: List<RealmMyLibrary?>, urls: ArrayList<String>) {
        for (library in lib) {
            val url = UrlUtils.getUrl(library)
            if (!FileUtils.checkFileExist(requireContext(), url) && !TextUtils.isEmpty(url)) {
                urls.add(url)
            }
        }
        if (urls.isNotEmpty()) {
            startDownload(urls)
        }
    }
    fun startDownloadWithAutoOpen(urls: ArrayList<String>, libraryToOpen: RealmMyLibrary? = null) {
        if (libraryToOpen != null) {
            pendingAutoOpenLibrary = libraryToOpen
            shouldAutoOpenAfterDownload = true
        }
        startDownload(urls)
    }
    override fun onDownloadComplete() {
        super.onDownloadComplete()
        if (shouldAutoOpenAfterDownload && pendingAutoOpenLibrary != null) {
            pendingAutoOpenLibrary?.let { library ->
                shouldAutoOpenAfterDownload = false
                pendingAutoOpenLibrary = null

                val isDownloaded = if (library.mediaType == "HTML") {
                    val directory = File(context?.getExternalFilesDir(null), "ole/${library.resourceId}")
                    val indexFile = File(directory, "index.html")
                    indexFile.exists()
                } else {
                    library.isResourceOffline() || FileUtils.checkFileExist(requireContext(), UrlUtils.getUrl(library))
                }

                if (isDownloaded) {
                    openResource(library)
                }
            }
        }
    }
    fun initRatingView(type: String?, id: String?, title: String?, listener: OnRatingChangeListener?) {
        timesRated = requireView().findViewById(R.id.times_rated)
        rating = requireView().findViewById(R.id.tv_rating)
        ratingBar = requireView().findViewById(R.id.rating_bar)
        ratingBar?.apply {
            setOnTouchListener { view, e: MotionEvent ->
                if (e.action == MotionEvent.ACTION_UP) {
                    view.performClick()
                }
                true
            }
            val userModel = profileDbHandler.userModel
            if (userModel?.isGuest() == false) {
                setOnClickListener {
                    homeItemClickListener?.showRatingDialog(type, id, title, listener)
                }
            }
        }
    }
    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnHomeItemClickListener) {
            homeItemClickListener = context
        }
    }

    private fun dismissProgressDialog() {
        try {
            if (prgDialog.isShowing()) {
                prgDialog.dismiss()
            }
        } catch (e: UninitializedPropertyAccessException) {
            e.printStackTrace()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
    fun openResource(items: RealmMyLibrary) {
        dismissProgressDialog()
        if (items.mediaType == "HTML") {
            openHtmlResource(items)
        } else {
            openNonHtmlResource(items)
        }
    }

    private fun openHtmlResource(items: RealmMyLibrary) {
        val directory = File(context?.getExternalFilesDir(null), "ole/${items.resourceId}")
        val indexFile = File(directory, "index.html")

        if (indexFile.exists()) {
            val intent = Intent(activity, WebViewActivity::class.java)
            intent.putExtra("RESOURCE_ID", items.id)
            intent.putExtra("LOCAL_ADDRESS", items.resourceLocalAddress)
            intent.putExtra("title", items.title)
            startActivity(intent)
            return
        }

        viewLifecycleOwner.lifecycleScope.launch {
            val resource = items.resourceId?.let { resourcesRepository.getLibraryItemByResourceId(it) }
            val downloadUrls = resource?.attachments
                ?.mapNotNull { attachment ->
                    attachment.name?.let { name ->
                        createAttachmentDir(items.resourceId, name)
                        UrlUtils.getUrl("${items.resourceId}", name)
                    }
                }
                ?.toCollection(ArrayList()) ?: arrayListOf()

            if (downloadUrls.isNotEmpty()) {
                startDownloadWithAutoOpen(downloadUrls, items)
            } else {
                val errorMessage = when {
                    resource == null -> getString(R.string.resource_not_found_in_database)
                    resource.attachments.isNullOrEmpty() -> getString(R.string.resource_has_no_attachments)
                    else -> getString(R.string.unable_to_download_resource)
                }
                Utilities.toast(activity, errorMessage)
            }
        }
    }

    private fun createAttachmentDir(resourceId: String?, name: String) {
        val baseDir = File(context?.getExternalFilesDir(null), "ole/$resourceId")
        val lastSlashIndex = name.lastIndexOf('/')
        if (lastSlashIndex > 0) {
            val dirPath = name.substring(0, lastSlashIndex)
            File(baseDir, dirPath).mkdirs()
        }
    }

    private fun openNonHtmlResource(items: RealmMyLibrary) {
        viewLifecycleOwner.lifecycleScope.launch {
            val matchingItems = items.resourceLocalAddress?.let {
                resourcesRepository.getLibraryItemsByLocalAddress(it)
            } ?: emptyList()

            val offlineItem = matchingItems.firstOrNull { it.isResourceOffline() }
            if (offlineItem != null) {
                ResourceOpener.openFileType(requireActivity(), offlineItem, "offline", profileDbHandler)
                return@launch
            }

            when {
                items.isResourceOffline() -> ResourceOpener.openFileType(
                    requireActivity(), items, "offline", profileDbHandler
                )
                FileUtils.getFileExtension(items.resourceLocalAddress) == "mp4" -> ResourceOpener.openFileType(
                    requireActivity(), items, "online", profileDbHandler
                )
                else -> {
                    val arrayList = arrayListOf(UrlUtils.getUrl(items))
                    startDownloadWithAutoOpen(arrayList, items)
                    profileDbHandler.setResourceOpenCount(items, KEY_RESOURCE_DOWNLOAD)
                }
            }
        }
    }

    private fun installApk(items: RealmMyLibrary) {
        if (BuildConfig.LITE) return
        currentLibrary = items
        val directory = File(requireContext().getExternalFilesDir(null).toString() + "/ole" + "/" + items.id)
        if (!directory.exists()) {
            if (!directory.mkdirs()) {
                throw RuntimeException("Failed to create directory: " + directory.absolutePath)
            }
        }
        val apkFile = items.resourceLocalAddress?.let { File(directory, it) }
        if (apkFile != null) {
            if (!apkFile.exists()) {
                Utilities.toast(activity,"APK file not found")
                return
            }
        }
        val uri = apkFile?.let {
            FileProvider.getUriForFile(requireContext(), "${requireContext().packageName}.fileprovider", it)
        }
        val intent = Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uri, "application/vnd.android.package-archive")
            flags = Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_ACTIVITY_NEW_TASK
        }
        if (intent.resolveActivity(requireActivity().packageManager) != null) {
            if (hasInstallPermission(requireContext())) {
                startActivity(intent)
            } else {
                requestInstallPermission()
            }
        } else {
            Utilities.toast(activity,"No app to handle the installation")
        }
    }

    private fun requestInstallPermission() {
        if (BuildConfig.LITE) return
        val intent = Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES)
        intent.data = ("package:" + requireContext().packageName).toUri()
        installApkLauncher?.launch(intent)
    }

    private fun openFileType(items: RealmMyLibrary, videoType: String) {
        dismissProgressDialog()
        ResourceOpener.openFileType(requireActivity(), items, videoType, profileDbHandler)
    }

    private fun showResourceList(downloadedResources: List<RealmMyLibrary>) {
        val builderSingle = AlertDialog.Builder(ContextThemeWrapper(requireActivity(), R.style.CustomAlertDialog))
        builderSingle.setTitle(getString(R.string.select_resource_to_open))
        val arrayAdapter: ArrayAdapter<RealmMyLibrary?> = object : ArrayAdapter<RealmMyLibrary?>(
            requireActivity(), android.R.layout.select_dialog_item, downloadedResources
        ) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                var view = convertView
                if (view == null) view = LayoutInflater.from(activity)
                    .inflate(android.R.layout.select_dialog_item, parent, false)
                val tv = view as TextView
                val library = getItem(position)
                tv.setCompoundDrawablesWithIntrinsicBounds(0, 0,
                    if (library?.isResourceOffline() == true) {
                        R.drawable.ic_eye
                    } else {
                        R.drawable.ic_download
                    }, 0)
                tv.setTextColor(context.getColor(R.color.daynight_textColor))
                tv.text = library?.title
                return tv
            }
        }
        builderSingle.setAdapter(arrayAdapter) { _: DialogInterface?, i: Int ->
            val library = arrayAdapter.getItem(i)
            library?.let { openResource(it) }
        }
        builderSingle.setNegativeButton(R.string.dismiss, null).show()
    }

    fun setOpenResourceButton(downloadedResources: List<RealmMyLibrary>?, btnOpen: Button) {
        if (downloadedResources.isNullOrEmpty()) {
            btnOpen.visibility = View.GONE
        } else {
            btnOpen.visibility = View.VISIBLE
            btnOpen.setOnClickListener {
                if (downloadedResources.size == 1) {
                    openResource(downloadedResources[0])
                } else {
                    showResourceList(downloadedResources)
                }
            }
        }
    }
    fun setResourceButton(resources: List<RealmMyLibrary>?, btnResources: Button) {
        if (resources.isNullOrEmpty()) {
            btnResources.visibility = View.GONE
        } else {
            btnResources.visibility = View.VISIBLE
            btnResources.text = getString(R.string.resources_size, resources.size)
            btnResources.setOnClickListener {
                if (resources.isNotEmpty()) {
                    showDownloadDialog(resources)
                }
            }
        }
    }

    open fun handleBackPressed() {
        NavigationHelper.popBackStack(parentFragmentManager)
    }

    override fun onPause() {
        super.onPause()
        dismissProgressDialog()
    }

    override fun onDestroy() {
        dismissProgressDialog()
        super.onDestroy()
    }
}
